### PR TITLE
feat(SF-979): Make requests after updating the URL

### PR DESCRIPTION
**Pull request by Vic-Dev**

### CODE REVIEW

File: src/core/actions/creators.ts [hunk 1]
```diff
@@ -43,7 +43,7 @@ namespace ActionCreators {
    * @param {Actions.Payload.Fetch.Override} options - An object with a request object for override.
    * @return {Actions.FetchProducts} - Action with `{ request }`.
    */
-  export function fetchProducts(options: Actions.Payload.Fetch.Override = {}): Actions.FetchProducts {
+  export function fetchProducts(options: Actions.Payload.Fetch.Override = { fetch: false }): Actions.FetchProducts {
```
Comment (Vic-Dev) [inline src/core/actions/creators.ts]: Remove `fetch: false`

File: src/core/actions/creators.ts [hunk 2]
```diff
@@ -53,7 +53,7 @@ namespace ActionCreators {
    * @return {Actions.FetchProductsWithoutHistory} - Action with `{ request }`.
    */
   // tslint:disable-next-line max-line-length
-  export function fetchProductsWithoutHistory(options: Actions.Payload.Fetch.Override = {}): Actions.FetchProductsWithoutHistory {
+  export function fetchProductsWithoutHistory(options: Actions.Payload.Fetch.Override = { fetch: false }): Actions.FetchProductsWithoutHistory {
```
Comment (Vic-Dev) [inline src/core/actions/creators.ts]: Remove `fetch: false`

File: src/core/actions/creators.ts [hunk 3]
```diff
@@ -63,8 +63,8 @@ namespace ActionCreators {
    * @return {Actions.FetchProductsWhenHydrated} - Action with `{ request }`.
    */
   // tslint:disable-next-line max-line-length
-  export function fetchProductsWhenHydrated(options: Actions.Payload.Fetch.Override = {}): Actions.FetchProductsWhenHydrated {
-    return createAction(Actions.FETCH_PRODUCTS_WHEN_HYDRATED, ActionCreators.fetchProducts(options));
+  export function fetchProductsWhenHydrated(options: Actions.Payload.Fetch.Override = { fetch: false }): Actions.FetchProductsWhenHydrated {
```
Comment (Vic-Dev) [inline src/core/actions/creators.ts]: Remove `fetch: false`

File: src/core/actions/creators.ts [hunk 4]
```diff
@@ -63,8 +63,8 @@ namespace ActionCreators {
...
+    return createAction(Actions.FETCH_PRODUCTS_WHEN_HYDRATED, <any>{ action: ActionCreators.fetchProducts(options), fetch: options.fetch });
```
Comment (Vic-Dev) [inline src/core/actions/creators.ts]: remove `fetch`

File: src/core/actions/creators.ts [hunk 5]
```diff
@@ -175,10 +175,10 @@ namespace ActionCreators {
    * @param  {string} id - The id for a specific product.
    * @return {Actions.FetchProductDetails} - Action with `{ id }`.
    */
-  export function fetchProductDetails(id: string): Actions.FetchProductDetails;
+  export function fetchProductDetails(id: string, fetch?: boolean): Actions.FetchProductDetails;
```
Comment (Vic-Dev) [inline src/core/actions/creators.ts]: remove `fetch`

File: src/core/actions/creators.ts [hunk 6]
```diff
@@ -175,10 +175,10 @@ namespace ActionCreators {
...
   // tslint:disable-next-line typedef
-  export function fetchProductDetails(options): Actions.FetchProductDetails {
-    const opts = typeof options === 'string' ? { id: options } : options;
+  export function fetchProductDetails(options, fetch = false): Actions.FetchProductDetails {
```
Comment (Vic-Dev) [inline src/core/actions/creators.ts]: remove `fetch`

File: src/core/actions/creators.ts [hunk 7]
```diff
@@ -220,7 +220,7 @@ namespace ActionCreators {
    * @return {Actions.FetchPastPurchaseProducts} - Action with `{ query, request }`.
    */
   // tslint:disable-next-line max-line-length
-  export function fetchPastPurchaseProducts(options: Actions.Payload.Fetch.PastPurchases = {}): Actions.FetchPastPurchaseProducts {
+  export function fetchPastPurchaseProducts(options: Actions.Payload.Fetch.PastPurchases = { fetch: false }): Actions.FetchPastPurchaseProducts {
```
Comment (Vic-Dev) [inline src/core/actions/creators.ts]: remove `fetch`

File: src/core/actions/creators.ts [hunk 8]
```diff
@@ -178,7 +178,7 @@ namespace ActionCreators {
   export function fetchProductDetails(id: string): Actions.FetchProductDetails;
   // tslint:disable-next-line typedef
   export function fetchProductDetails(options): Actions.FetchProductDetails {
-    const opts = typeof options === 'string' ? { id: options } : options;
+    const opts = typeof options === 'string' ? { id: options } : { ...options };
```
Comment (Vic-Dev) [inline src/core/actions/creators.ts]: just return `options`

File: src/core/actions/creators.ts [hunk 9]
```diff
@@ -535,13 +535,13 @@ namespace ActionCreators {
...
-   * Sets the details product in the store, doing some additional emits and state changes.
+   * Sets the details product in the store with an optional template.
    * @param  {Record | Store.Product}         product - The product to use as the details product.
    * @param  {Template}         template - The template to use as the details template.
-   * @return {Actions.SetDetails}         - Action with details.
+   * @return {Actions.UpdateDetails}         - Action with details.
    */
-  export function setDetails(product: Record | Store.Product, template?: Template): Actions.SetDetails {
-    return createAction(Actions.SET_DETAILS, { data: product, template: SearchAdapter.extractTemplate(template) });
+  export function setDetails(product: Record | Store.Product, template?: Template): Actions.UpdateDetails {
```
Comment (Vic-Dev) [inline src/core/actions/creators.ts]: If `setDetails` just returns an `updateDetails` action we should just deprecate/remove `setDetails`, make sure clients using it fix their code, and fix where we use `setDetails`

File: src/core/adapters/request.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+import { MAX_RECORDS } from './search';
+
+namespace Adapter {
+  export const clampPageSize = (page: number, pageSize: number): number =>
+    Math.min(pageSize, MAX_RECORDS - Adapter.extractSkip(page, pageSize));
+
+  export const extractSkip = (page: number, pageSize: number): number =>
+    (page - 1) * pageSize;
+
+  // TODO: add Sort return value when api-javascript goes out
```
Comment (Vic-Dev) [inline src/core/adapters/request.ts]: Do this

File: src/core/events.ts [hunk 11]
```diff
@@ -148,6 +148,16 @@ namespace Events {
    * Returns the searchId.
    */
   export const SEARCH_CHANGED = 'search_changed';
+  /**
+   * Triggered when detailsId changes. Occurs whenever a new search is fired.
```
Comment (Vic-Dev) [inline src/core/events.ts]: `new details search`

File: src/core/reducers/session.ts [hunk 12]
```diff
@@ -18,6 +18,12 @@ export default function updateSession(state: State = {}, action: Action): State
         if ('searchId' in action.meta) {
```
Comment (Vic-Dev) [inline src/core/reducers/session.ts]: these can all just use a function that's something like `updateSection` where it updates the key given

File: src/core/sagas/index.ts [hunk 13]
```diff
@@ -6,6 +6,7 @@ import productDetails from './product-details';
 import products from './products';
 import recommendations from './recommendations';
 import refinements from './refinements';
+//import router from './router';
```
[the remark this answers is not in the document]
Comment (Vic-Dev) [reply src/core/sagas/index.ts]: remove

File: src/core/sagas/index.ts [hunk 14]
```diff
@@ -15,7 +16,8 @@ export const SAGA_CREATORS = [
   productDetails,
   products,
   recommendations,
-  refinements
+  refinements,
+//  router
```
Comment (Vic-Dev) [inline src/core/sagas/index.ts]: remove

File: src/core/actions/creators.ts [hunk 15]
```diff
@@ -535,13 +535,13 @@ namespace ActionCreators {
   }
```
Comment (Vic-Dev) [inline src/core/actions/creators.ts]: `updateDetails` should handle templates

File: src/core/store/middleware.ts [hunk 16]
```diff
@@ -187,9 +188,9 @@ export namespace Middleware {
     const composeEnhancers = global && global['__REDUX_DEVTOOLS_EXTENSION_COMPOSE__'] || compose;
 
     return composeEnhancers(
-      applyMiddleware(...normalizingMiddleware, saveStateAnalyzer, pastPurchaseProductAnalyzer),
+      applyMiddleware(...normalizingMiddleware, saveStateAnalyzer),
```
Comment (Vic-Dev) [inline src/core/store/middleware.ts]: Maybe try to get rid of these last steps bc shouldn't be necessary...

File: src/core/store/middleware.ts [hunk 17]
```diff
@@ -187,9 +188,9 @@ export namespace Middleware {
...
       applyMiddleware(...middleware),
-      applyMiddleware(...normalizingMiddleware, Middleware.validator),
+      applyMiddleware(...normalizingMiddleware),
```
Comment (Vic-Dev) [inline src/core/store/middleware.ts]: Maybe try to get rid of this first step bc shouldn't be necessary...

File: src/flux-capacitor.ts [hunk 18]
```diff
@@ -100,9 +100,7 @@ class FluxCapacitor extends EventEmitter {
   }
 
   detailsWithRouting(product: Store.Product | Record) {
-    return this.config.details.alwaysFetch
-      ? this.store.dispatch(this.actions.fetchProductDetails(SearchAdapter.extractAllMeta(product).id))
-      : this.store.dispatch(this.actions.setDetails(product));
+    this.store.dispatch(this.actions.setDetails(product));
```
Comment (Vic-Dev) [inline src/flux-capacitor.ts]: use `updateDetails` instead and pass the right thing

File: src/flux-capacitor.ts [hunk 19]
```diff
@@ -100,9 +100,7 @@ class FluxCapacitor extends EventEmitter {
...
   }
```
Comment (Vic-Dev) [inline src/flux-capacitor.ts]: remove `alwaysFetch` from configuration bc details should always always fetch now

File: src/core/store/middleware.ts (hunk 16 above)
Comment (Vic-Dev) [reply src/core/store/middleware.ts]: do later

File: src/core/store/middleware.ts (hunk 17 above)
Comment (Vic-Dev) [reply src/core/store/middleware.ts]: do later

File: test/unit/core/reducers/data/details.ts [hunk 20]
```diff
@@ -17,18 +17,22 @@ suite('details', ({ expect }) => {
...
+        template: {
+          name: template.name,
+          rule: template.ruleName,
+          zones: template.zones
+        }
       };
 
       const reducer = details(state, {
         type: Actions.UPDATE_DETAILS,
-        payload: { data: product2, template }
```
Comment (jmbtutor) [inline test/unit/core/reducers/data/details.ts]: Remove extra space.

File: test/unit/core/sagas/recommendations.ts [hunk 21]
```diff
@@ -343,24 +346,25 @@ suite('recommendations saga', ({ expect, spy, stub }) => {
...
+          updatePastPurchasePageSize(),
           receivePastPurchasePage(),
           receivePastPurchaseCurrentRecordCount(),
           receivePastPurchaseProducts(),
         ]));
         task.next();
 
         expect(augmentProducts).to.be.calledWithExactly(productData);
-        expect(receivePastPurchasePage).to.be.calledWithExactly(productData, currentPage);
+        expect(receivePastPurchasePage).to.be.calledWithExactly(productData, currentPage, 30);
```
Comment (jmbtutor) [inline test/unit/core/sagas/recommendations.ts]: Extract `30` to a variable (`pageSize`?)

File: src/core/reducers/session.ts [hunk 22]
```diff
@@ -13,10 +13,16 @@ export default function updateSession(state: State = {}, action: Action): State
...
-          state = updateSearchId(state, action.meta);
+          state = updateSection(state, action.meta, 'searchId');
+        }
+        if ('detailsId' in action.meta) {
+          state = updateSection(state, action.meta, 'detailsId');
+        }
+        if ('pastPurchaseId' in action.meta) {
+          state = updateSection(state, action.meta, 'pastPurchaseId');
         }
```
[the remark this answers is not in the document]
Comment (Vic-Dev) [reply src/core/reducers/session.ts]: sure! don't forget `'searchId'`

File: src/core/reducers/session.ts [hunk 23]
```diff
@@ -45,11 +51,8 @@ export const updateSecuredPayload = (state, securedPayload: Configuration.Recomm
...
-  ({ ...state, recallId });
-
-export const updateSearchId = (state: State, { searchId }: Actions.Metadata) =>
-  ({ ...state, searchId });
+  export const updateOrigin = (state: State, { tag: origin }: Actions.Metadata) =>
+    ({ ...state, origin });
 
-export const updateOrigin = (state: State, { tag: origin }: Actions.Metadata) =>
-  ({ ...state, origin });
+export const updateSection = (state: State, value: Actions.Metadata, section: string) =>
```
[the remark this answers is not in the document]
Comment (Vic-Dev) [reply src/core/reducers/session.ts]: makes snes

File: src/core/reducers/data/details.ts [hunk 24]
```diff
@@ -1,17 +1,20 @@
...
+    case Actions.UPDATE_DETAILS:
+    case Actions.RECEIVE_DETAILS:
+      return update(state, action.payload);
     default: return state;
   }
 }
 
 export const update = (state: State, { data, template }: Store.Details) =>
-  ({ ...state, data, template });
+  ({ ...state, data, template: SearchAdapter.extractTemplate(<any>template) });
```
[the remark this answers is not in the document]
Comment (Vic-Dev) [reply src/core/reducers/data/details.ts]: yes

**Pull request merged**